File: AlgorithmRise/extra/LeetCode/Algorithms/687. Longest Univalue Path/draft-1.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int max = -1;
void traverseTree(struct TreeNode* pre, struct TreeNode* p, bool is_continue, int path){

    if(!p){
        return;
    }

    if(!pre){

        is_continue = true;
    }else{

        if(pre->val != p->val){
            is_continue = false;
        }else{
            is_continue = true;
        }

        if(is_continue){
            ++path;
            max = path > max ? path : max;
        }
    }

    if(p->left){
        traverseTree(p, p->left, is_continue, path);
    }

    if(p->right){
        traverseTree(p, p->right, is_continue, path);
    }
}

int longestUnivaluePath(struct TreeNode* root){

    max = -1;
    traverseTree(NULL, root, true, 0);

    return max;
}

File: AlgorithmRise/extra/LeetCode/Algorithms/101. Symmetric Tree/draft-1.go
package  SymmetricTree

var nodes [][] int

func reverse(a []int) []int{
    b := make([]int, len(a))
    for i := len(a)/2-1; i >= 0; i-- {
	    opp := len(a)-1-i
	    b[i], b[opp] = a[opp], a[i]
    }
    fmt.Println(a)
    fmt.Println(b)
    return b
}

func max(vals ...int) int {
    max := 0
    for _, val := range vals {
        if val >= max{
            max = val
        }
    }
    return max
}

func traverseTree(node *TreeNode, level int) int {

	if nil == node {
		return level
	}

    nodes[level] = append(nodes[level], node.Val)

    var level_l, level_r int
	if nil != node.Left {
		level_l = traverseTree(node.Left, level+1)
	}

	if nil != node.Right {
		level_r = traverseTree(node.Right, level+1)
	}

    // 返回 Max_level
    return max(level, level_l, level_r)
}

func sliceEqual(a, b []int) bool {
    if len(a) != len(b) {
        return false
    }

    if (a == nil) != (b == nil) {
        return false
    }

    for i, v := range a {
        if v != b[i] {
            return false
        }
    }

    return true
}

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSymmetric(root *TreeNode) bool {

    nodes = make([][]int, 1005)
    max_level := traverseTree(root, 0)

    for i:= 0; i<= max_level; i++ {

        arr := nodes[i]
        _arr := reverse(arr)
        if false == sliceEqual(arr, _arr) {
            return false;
        }
    }
    return true
}